### DISCUSSION TITLE
Noise below the mixed layer when using `AnisotropicMinimumDissipation`

### DISCUSSION THREAD

**Discussion by jhdong2016** [CliMA/Oceananigans.jl#3561] [Computational science!]
Hi all, I am simulating an ocean surface mixed layer. However, I always get lots of noise in the ocean interior below the mixed layer. I  have applied a Gaussian-mask sponge layer near the bottom. But the noise is still there. Below are two figures of the vertical velocity and buoyancy perturbations.  Any suggestions?
![b](https://github.com/CliMA/Oceananigans.jl/assets/46232007/f160c0b6-df59-4d8a-89a2-b6de279a6e23)
![w](https://github.com/CliMA/Oceananigans.jl/assets/46232007/1d42358b-c512-41c4-bffb-035a6a169b2b)

Here are the codes for the sponge.

ps. the whole depth is Lz = 150 m and the mixed layer is about 80m.

######========================================
damping_rate = 1/1800 # relax fields on a 100 second time-scale
bottom_mask = GaussianMask{:z}(center=-grid.Lz, width=grid.Lz/3)
uvw_sponge = Relaxation(rate=damping_rate, mask=bottom_mask)

model = NonhydrostaticModel(; grid, coriolis,
                            advection = WENO(),
                            timestepper = :RungeKutta3,
                            tracers = :b,
                            buoyancy = BuoyancyTracer(),
                            closure = AnisotropicMinimumDissipation(),
                            stokes_drift = UniformStokesDrift(∂z_vˢ=∂z_vˢ),
                            boundary_conditions = (v=v_boundary_conditions, b=b_boundary_conditions),
                            forcing=(u=uvw_sponge, v=uvw_sponge, w=uvw_sponge))
#####============================================

**Comment by navidcy** [comment]
Hi @jhdong2016, thanks for your interest in Oceananigans! 

Personally I find it very difficult to help out without something that I can run and reproduce your results. Could you post your script somewhere also for making the plots. Something I can copy/paste in the REPL or just write `include("jhdong_script.jl")`?

It's often useful to post the version of Oceananigans you are using (and the rest of dependencies). E.g., post the output of `using Pkg; Pkg.status()`.

Do I understand correctly or is it my impression that you have depth in the horizontal axis?

**Comment by jhdong2016** [comment]
THanks for the reply. Here are the files that may help.

[si_langmuir_case0.8m.jl.txt](https://github.com/CliMA/Oceananigans.jl/files/15108771/si_langmuir_case0.8m.jl.txt)

<img width="522" alt="image" src="https://github.com/CliMA/Oceananigans.jl/assets/46232007/8463a044-59fa-4a55-a1e9-0b03ff2a3353">

**Comment by navidcy** [comment]
Thanks!

You seem to be using a very very old version of Oceananigans; v0.79.0 (from 2022).
Could you try to update to the latest version and see if the issue(s) remain?

**Comment by jhdong2016** [comment]
Thanks for the suggest! Let me update it and see if the issue is fixed.

**Comment by glwagner** [comment]
This is common to observe with `AnisotropicMinimumDissipation`. If you don't need the dissipation rate or eddy diffusivities explicitly, you could try using `closure=nothing` and `advection=WENO(order=9)` instead. This might solve your noise problem and also give you a higher effective resolution.

Can we convert this to a discussion? I don't think this is a bug in the code.

**Reply by glwagner** [reply]
@tomchor might have more to say

**Comment by glwagner** [comment]
And yes more generally it's best if you include your whole script so that the simulation can be reproduced. In particular here we need the grid, initial condition, and time step to fully grasp what's doing on.

**Comment by jhdong2016** [comment]
> This is common to observe with `AnisotropicMinimumDissipation`. If you don't need the dissipation rate or eddy diffusivities explicitly, you could try using `closure=nothing` and `advection=WENO(order=9)` instead. This might solve your noise problem and also give you a higher effective resolution.
> 
> Can we convert this to a discussion? I don't think this is a bug in the code.

Please convert it into a discussion. And I am updating the version and will try your suggestion! Thanks!

**Comment by glwagner** [comment]
I'd like to revisit this. @jhdong2016 can you provide your entire script? Alternatively, @tomchor do you have a script that illustrates the noise issue? I would like to see if there is a simple fix.

**Reply by tomchor** [reply]
The code [here](https://github.com/tomchor/CSI_insights/blob/master/surfjet_np.jl), which is the code I used for the simulations in [this publication](https://journals.ametsoc.org/view/journals/phoc/52/10/JPO-D-21-0259.1.xml), reproduces the issue when AMD is used. That said, the code uses a very old version of Oceananigans and is very complex with all the bells and whistles I  needed for the systematic exploration. It might be easier to just read Section 3 of the paper and reproduce that simulation from scratch.

That said, I remember I was able to reproduce the noise with simpler simulations, I just don't remember more details, unfortunately. @johnryantaylor may have ideas on how to reproduce this issue more simply.

**Reply by tomchor** [reply]
Okay I spent some time today on this and came up with a small working example. I'm starting with a doubly-periodic box with a constant (strong) stratification and some noise:

```julia
using Oceananigans
using Oceananigans.Units

grid = RectilinearGrid(GPU(), size=(50, 50, 20), extent = (500, 500, 50))
@info grid

N²∞ = 4e-5
@inline bᵢ(x, y, z) = N²∞ * z

@info "Creating model"
model = NonhydrostaticModel(; grid,
                            buoyancy = BuoyancyTracer(), tracers = :b,
                            closure = AnisotropicMinimumDissipation(),
                            )
@info "" model

grid_noise(x, y, z) = 0.002 * rand()

set!(model, w = grid_noise, b = bᵢ)

u, v, w = model.velocities
b = model.tracers.b

CellCenter = (Center, Center, Center)
outputs = Dict{Any, Any}(:u => (@at CellCenter u),
                         :v => (@at CellCenter v),
                         :w => (@at CellCenter w),
                         :b => b)

simulation = Simulation(model, Δt = 0.5 * minimum_zspacing(grid) / maximum(abs, w), stop_time = 10hours)

wizard = TimeStepWizard(max_change=1.01, cfl=0.8, max_Δt=0.3/√N²∞)
simulation.callbacks[:wizard] = Callback(wizard, IterationInterval(10))

simulation.output_writers[:nc] = ow = NetCDFOutputWriter(model, outputs; filename = "AMD-mwe.nc",
                                                         schedule = TimeInterval(20minutes),
                                                         overwrite_existing = true, indices = (1, :, :))
run!(simulation)


fpath_iyz = simulation.output_writers[:nc].filepath
using NCDatasets
ds = NCDataset(fpath_iyz, "r")

axis_kwargs = (xlabel = "y (m)", ylabel = "Depth (m)", width = 300, height=100,)

using CairoMakie
n = Observable(1)

u = @lift ds["u"][1, :, :, $n]
v = @lift ds["v"][1, :, :, $n]
w = @lift ds["w"][1, :, :, $n]
b = @lift ds["b"][1, :, :, $n]
yc, zc = ds["yC"], ds["zC"]

fig = Figure()

u_lims = (-0.01, 0.01)
ax_u = Axis(fig[1, 1]; title = "u", axis_kwargs...)
hm_u = heatmap!(ax_u, yc, zc, u, colorrange = u_lims, colormap = :balance)
Colorbar(fig[1, 2], hm_u; label = "m s⁻¹")
ct_b = contour!(ax_u, yc, zc, b, levels=10, color=:black)

ax_w = Axis(fig[2, 1]; title = "w", axis_kwargs...)
hm_w = heatmap!(ax_w, yc, zc, w, colorrange = u_lims, colormap = :balance)
Colorbar(fig[2, 2], hm_w; label = "m/s")
ct_b = contour!(ax_w, yc, zc, b, levels=10, color=:black)

n_times = length(ds["time"])
frames = 1:n_times
resize_to_layout!(fig) # Resize figure after everything is done to it, but before recording
CairoMakie.record(fig, "$(@__DIR__)/$fpath_iyz.mp4", frames, framerate=14) do frame
    @info "Plotting time $frame of $(n_times)..."
    n[] = frame
end
close(ds)
```

The output of the code above is:

https://github.com/user-attachments/assets/5c96ec10-2aae-405b-8ccd-ae7c528c4ed4

where you can see that the initial noise gets amplified. The issue does appear to be somewhat resolution-dependent though. The simulation above has Δz=2.5m. If I get it down to 2m (by decreasing the size of the box) the output is then:

https://github.com/user-attachments/assets/91ac892b-4556-4a86-ad1b-92cc26d6f69f

Which still amplifies initial noise, but it appear to be less severe.

Also as a sanity check, the same simulation without AMD produces:

https://github.com/user-attachments/assets/f6c66adf-0972-444e-9a0a-17b06daa2779

So it really seems to be AMD that's causing this.

**Reply by ali-ramadhan** [reply]
Just thinking out loud here but should AMD be used with some background scalar diffusivity? It could probably diffuse the noise (as desired?).

I remember that in ancient versions of Oceananigans you could set a background diffusivity and viscosity for AMD, e.g. from v0.20.0:

https://github.com/CliMA/Oceananigans.jl/blob/88a4f8f08d976bb0029ea337202bf94fe292374d/src/TurbulenceClosures/turbulence_closure_implementations/verstappen_anisotropic_minimum_dissipation.jl#L30-L33

and the defaults were the molecular values

https://github.com/CliMA/Oceananigans.jl/blob/88a4f8f08d976bb0029ea337202bf94fe292374d/src/TurbulenceClosures/TurbulenceClosures.jl#L45-L53

I don't remember if background diffusivities were part of AMD or a necessary hack as Oceananigans didn't have tupled turbulence closures back then.

**Reply by glwagner** [reply]
Thanks @tomchor ! Is adaptive time-stepping crucial? I'm wondering if using a short time-step produces the same results.

Is the total energy increasing in these simulations? It seems like it has to be.

**Reply by glwagner** [reply]
I tried to simplify the script but I'm having trouble getting anything interesting. Not sure what's going wrong.

```julia
using Oceananigans
using Oceananigans.Units
using Random

function simulate_noisy_stratification(closure, filename, stop_time)
    Random.seed!(42)
    Nx = Ny = Nz = 32
    x = y = z = (0, 10Nx)
    grid = RectilinearGrid(size=(Nx, Ny, Nz); x, y, z)
    model = NonhydrostaticModel(; grid, closure, tracers=:b, buoyancy=BuoyancyTracer())
    bᵢ(x, y, z) = 1e-4 * z
    noise(x, y, z) = 1e-12 * rand()
    set!(model, w=noise, b=bᵢ)
    simulation = Simulation(model, Δt=100.0; stop_time)
    progress(sim) = @info string(iteration(simulation), " at ", prettytime(simulation))
    add_callback!(simulation, progress, IterationInterval(100))

    simulation.output_writers[:nc] = JLD2OutputWriter(model, model.velocities; filename,
                                                      schedule = TimeInterval(20minutes),
                                                      overwrite_existing = true,
                                                      indices = (1, :, :))
    run!(simulation)

    return nothing
end

using GLMakie

amd = AnisotropicMinimumDissipation()
simulate_noisy_stratification(amd, "amd.jld2", 4 * 24hours)
simulate_noisy_stratification(nothing, "nothing.jld2", 4 * 24hours)

uat = FieldTimeSeries("amd.jld2", "u")
unt = FieldTimeSeries("nothing.jld2", "u")
Nt = length(uat)

fig = Figure(size=(1200, 600))
axua = Axis(fig[1, 1], aspect=1)
axun = Axis(fig[1, 2], aspect=1)
slider = Slider(fig[2, 1:2], range=1:Nt, startvalue=1)
n = slider.value
uan = @lift uat[$n]
unn = @lift unt[$n]
ulim = 2 * maximum(abs, uat[1])
heatmap!(axua, uan, colorrange=(-ulim, ulim))
heatmap!(axun, unn, colorrange=(-ulim, ulim))
display(fig)
```

**Reply by tomchor** [reply]
> Thanks @tomchor ! Is adaptive time-stepping crucial? I'm wondering if using a short time-step produces the same results.

I haven't tested without it, but I'd guess it's not necessary.

> Is the total energy increasing in these simulations? It seems like it has to be.

I also haven't measured it, but I think has to be given how visually striking the noise amplification is!

About your MWE. I can't see anything that would obviously kill the effect. I'd suggest changing one thing at a time from my working example.

**Reply by jhdong2016** [reply]
Wow, sorry that I miss so much information. I updated the version and used the center2ndorder advection scheme, and there seems to be no noise now. My feeling is that the choice of the advection scheme is really important.

**Reply by glwagner** [reply]
> I'd suggest changing one thing at a time from my working example.

They are already identical except for parameter choices and adaptive time stepping.

**Reply by glwagner** [reply]
Also just as a PS --- powers of 2 are fastest for NonhydrostaticModel due to the FFT.

Multiples of 10 are ok, because cufft supports prime factorizations of 2 and 5, for sure. However, power of 2 are still faster. I think (but am not sure) that powers of 3 are faster than powers of 5. So it is nice to follow a convention that uses just powers of 2 and 3.

From an aesthetic point of view, powers of 2 are actually nice. It's the least arbitrary basis. (It's believed we use base 10 because we have 10 fingers, which is also kind of cool but still, a bit more arbitrary.)

Personally I think communication is enhanced and mental noise is reduced if we always use powers of 2 and 3 for resolution, plus grid spacings that are simple (ie do not combine "Lx = 100" with "Nx = 64" --- use "Lx = 128" instead).

You can read more here: https://docs.nvidia.com/cuda/cufft/

**Reply by glwagner** [reply]
The other parts of the "MWE" that I found to be non-minimal were the interpolation of output to cell centers. This is extraneous to the AMD issue (hopefully)... but complicates the script since I spent some time reading it. I also think MWEs should use JLD2 at the moment, since NetCDF still (and unfortunately) does not support FieldTimeSeries which complicates visualization. I think the "minimal" part of MWE should really be taken seriously.

**Reply by tomchor** [reply]
I agree with you. If you read my post I called it a "small working example", not a minimal one ;)

I interpolated the outputs to make plotting easier with NetCDF (locations are the same) and I was using NetCDF because I can explore the solutions faster with xarray. I of course could have spent the time to trim down those things before posting, but I was running out of time and I wanted to make sure I posted it here, rather than have the example forgotten in my laptop (which is what happened the last time I investigated this issue).

**Reply by glwagner** [reply]
If FieldTimeSeries is missing useful features from xarray then we should have an issue open about it I think

**Reply by tomchor** [reply]
Was playing around with your example and have a working version with an animation at the end using JLD2:

```julia
using Oceananigans
using Oceananigans.Units
using Random: seed!

function simulate_noisy_stratification(closure, filename; stop_time = 5hours, Nz = 32, N² = 4e-5, Ξ = 0.002, β = 10)
    grid = RectilinearGrid(CPU(), size=(Nz, Nz, Nz), extent = (10Nz, 10Nz, 2Nz))
    model = @show NonhydrostaticModel(; grid, closure, buoyancy = BuoyancyTracer(), tracers = :b)

    seed!(42)
    grid_noise(x, y, z) = y > 5Nz ? Ξ/10 * rand() : Ξ * rand()
    bᵢ(x, y, z) = z > -Nz ? N² * z : N² * ((z + Nz)/β - Nz)
    set!(model, w = grid_noise, b = bᵢ)

    simulation = Simulation(model; Δt = 0.1 * minimum_zspacing(grid) / maximum(abs, model.velocities.w), stop_time)
    conjure_time_step_wizard!(simulation, IterationInterval(5), cfl = 0.2, max_Δt = 0.3/√N²)
    simulation.output_writers[:slice] = JLD2OutputWriter(model, fields(model); filename,
                                                         schedule = TimeInterval(10minutes),
                                                         overwrite_existing = true, indices = (1, :, :))
    run!(simulation)
end

simulate_noisy_stratification(AnisotropicMinimumDissipation(), "amd.jld2", stop_time=10hours)
simulate_noisy_stratification(nothing, "nothing.jld2", stop_time=10hours)

uat = FieldTimeSeries("amd.jld2", "u")
unt = FieldTimeSeries("nothing.jld2", "u")

using CairoMakie
fig = Figure(size=(1200, 600))
axua = Axis(fig[1, 1], aspect=1, title = "closure=AMD")
axun = Axis(fig[1, 2], aspect=1, title = "closure=nothing")

n = Observable(1)
uan = @lift uat[$n]
unn = @lift unt[$n]
ulim = 2 * maximum(abs, unt)

hma = heatmap!(axua, uan, colorrange = (-ulim, ulim), colormap = :balance)
hma = heatmap!(axun, unn, colorrange = (-ulim, ulim), colormap = :balance)
Colorbar(fig[2, 1], hma; label = "m s⁻¹", vertical = false)
Colorbar(fig[2, 2], hma; label = "m s⁻¹", vertical = false)

frames = 1:length(uat.times)
record(fig, "amd_mwe.mp4", frames, framerate=14) do frame
    @info "Plotting time $frame of $(length(frames))..."
    n[] = frame
end

```
The initialization of the above script happens with a stratification that's 10x stronger at the top half of the domain than at the bottom half. And the grid noise amplitude is 10x stronger on the left half of the domain than on the right half. This let's us test four combinations (one per quadrant). This is what this script produces:

https://github.com/user-attachments/assets/4b247b56-6e6e-4289-bea0-77d521108a66

So it seems like the growth rate of the instabilities is associated with both the initial noise value and the stratification.

**Reply by tomchor** [reply]
There seems to be a point where the growth rate is damped enough to kill the noise growth. For example, if I set `N² = 1e-8; Ξ = 0.002` this is what I get:

https://github.com/user-attachments/assets/dd260991-dce3-4365-b3ea-64396c8d550c

So maybe we need to adjust a coefficient based on `w` and `dbdz` or something.

**Reply by glwagner** [reply]
Nice thinking, this is revealing. The issue is with the part of AMD that models the tracer diffusivity, right? Not the eddy viscosity? Is there any way to test that?

**Reply by tomchor** [reply]
I _think_ it has to do with the diffusivity. But I did try fixing the Prandtl number (therefore avoiding an independent calculation of diffusivity) and the issue persisted, so maybe it's more complex than that.

**Reply by glwagner** [reply]
Ah doesn't that prove that it's not the diffusivity? If this persists with WENO then you can try to do something silly like turn the viscosity off completely (in the source) but keep diffusivity to check (using WENO so things stay stable).

**Reply by tomchor** [reply]
Yeah, that's a good idea. I think it's indeed time to start playing around with the source code. I won't have time to work on this for a while, but I'll keep that in mind for my next iteration :)

Feel free to try this in the meantime if you want though!
